File: HomeWork1/Hw1Tests/CalculatorTests.cs
using Hw1.Calculator;
using Xunit;

namespace Hw1Tests
{
    public class CalculatorTests
    {
        [Theory]
        [InlineData(5, CalculatorOperation.Plus, 7, 12)]
        [InlineData(-5, CalculatorOperation.Plus, 7, 2)]
        [InlineData(-5, CalculatorOperation.Plus, -7, -12)]
        [InlineData(5, CalculatorOperation.Plus, -7, -2)]
        public void Calculate_Plus_WillReturnCorrectResult(int val1, CalculatorOperation operation, int val2, int expected)
        {
            var result = Calculator.Calculate(val1, operation, val2);
            Assert.Equal(expected, result);
        }
        
        [Theory]
        [InlineData(5, CalculatorOperation.Minus, 7, -2)]
        [InlineData(-5, CalculatorOperation.Minus, 7, -12)]
        [InlineData(-5, CalculatorOperation.Minus, -7, 2)]
        [InlineData(5, CalculatorOperation.Minus, -7, 12)]
        public void Calculate_Minus_WillReturnCorrectResult(int val1, CalculatorOperation operation, int val2, int expected)
        {
            var result = Calculator.Calculate(val1, operation, val2);
            Assert.Equal(expected, result);
        }
        
        [Theory]
        [InlineData(5, CalculatorOperation.Multiply, 7, 35)]
        [InlineData(-5, CalculatorOperation.Multiply, 7, -35)]
        [InlineData(-5, CalculatorOperation.Multiply, -7, 35)]
        [InlineData(5, CalculatorOperation.Multiply, -7, -35)]
        public void Calculate_Multiply_WillReturnCorrectResult(int val1, CalculatorOperation operation, int val2, int expected)
        {
            var result = Calculator.Calculate(val1, operation, val2);
            Assert.Equal(expected, result);
        }

        [Theory]
        [InlineData(10, CalculatorOperation.Divide, 5, 2)]
        [InlineData(-10, CalculatorOperation.Divide, 5, -2)]
        [InlineData(-10, CalculatorOperation.Divide, -5, 2)]
        [InlineData(10, CalculatorOperation.Divide, -5, -2)]
        [InlineData(7, CalculatorOperation.Divide, 2, 3.5)]
        public void Calculate_Divide_WillReturnCorrectResult(int val1, CalculatorOperation operation, int val2, double expected)
        {
            var result = Calculator.Calculate(val1, operation, val2);
            Assert.Equal(expected, result);
        }
    }
}

File: HomeWork5/Hw5Tests/Calculator.Tests.fs
module Homework5.Tests.Calculator_Tests
open System
open Homework5.Calculator
open Homework5.ErrorType
open Xunit
open Homework5
open FailTest

let checkCorrectCalculate val1 operation val2 expected =
    let result = Calculate(val1,operation,val2)
    match result with
    | Ok result' -> Assert.Equal(expected,result')
    | Error e -> fail
    
[<Fact>]
let ``calculate_CorrectArguments_CorrectResultReturned``() =
    checkCorrectCalculate 220.m Operation.Plus 8.m 228.m 
    checkCorrectCalculate 1500.m Operation.Minus 2.m 1498.m
    checkCorrectCalculate 22.m Operation.Multiply 3.m 66.m
    checkCorrectCalculate 25.m Operation.Divide 4.m 6.25m


[<Fact>]
let ``calculate_DivideByZero_ErrorReturned``() =
    let result = Calculate (23.m, Operation.Divide, 0.m)
    match result with
    | Ok result' -> fail
    | Error e -> Assert.Equal(DivideByZero,e)